### PR TITLE
[BEAM-14347] Update docs to prefer generic registration functions

### PR DESCRIPTION
**Pull request by damccorm**
Now that we've proven out the effectiveness of the generic registration functions in https://github.com/apache/beam/pull/17613, we should update the docs to recommend that approach. This is the last piece of work for the generic registration to be done for now.

✨ [RENDERED](http://apache-beam-website-pull-requests.storage.googleapis.com/17635/documentation/programming-guide/index.html) ✨ 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: website/www/site/content/en/documentation/programming-guide.md [hunk 1]
```diff
@@ -432,12 +432,6 @@ around to distributed workers). The Beam SDKs provide a data encoding mechanism
 that includes built-in encoding for commonly-used types as well as support for
 specifying custom encodings as needed.
 
-{{< paragraph class="language-go" >}}
```
Comment (jrmccluskey) [inline website/www/site/content/en/documentation/programming-guide.md:435]: Does the generics approach also remove the need to register structs?

**Review by jrmccluskey — APPROVED** [review]

File: website/www/site/content/en/documentation/programming-guide.md (hunk 1 above)
Comment (damccorm) [reply website/www/site/content/en/documentation/programming-guide.md:435]: Yeah, all the structs will be inferred and registered automatically

**Comment by damccorm** [comment]
I'm actually going to override the bot and send this to Robert since he's been looking at the other registration stuff

R: @lostluck

**Comment by lostluck** [comment]
Consider splitting the CHANGES.md messaging out to it's own PR. 

 Since 2.40 doesn't release until July, the generics registration shouldn't yet be in the Programming Guide, when they can't be used for 6-8 weeks.

**Comment by damccorm** [comment]
Oh, forgot that the website doesn't get updated with releases. I opened https://github.com/apache/beam/pull/17643 and will close here for now

File: sdks/go/examples/snippets/04transforms.go [hunk 2]
```diff
@@ -185,6 +190,8 @@ func formatCoGBKResults(key string, emailIter, phoneIter func(*string) bool) str
 
 func init() {
 	beam.RegisterFunction(formatCoGBKResults)
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/examples/snippets/04transforms.go:192]: Whoops missed publishing these before you closed it.

WRT documentation changes, we likely do want to hold on making the beam package GoDoc mention these until we're unblocked from the import process.

But the other example code (everything in beam/examples) and integration test code can certainly make use of it before then. Doesn't need to be all at once, we can divide it amongst the team so folks can get used to it.

**Pull request closed without merging**